File: problems/1488_avoid_flood_in_the_city/1488_avoid_flood_in_the_city.py
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        n = len(rains)
        ans = [-1] * n
        dryDays = SortedList()
        lastRain = {}
        
        for i, lake in enumerate(rains):
            if lake > 0:
                if lake in lastRain:
                    idx = dryDays.bisect_right(lastRain[lake])
                    if idx == len(dryDays):
                        return []
                    dry = dryDays[idx]
                    ans[dry] = lake
                    dryDays.remove(dry)
                lastRain[lake] = i
            else:
                dryDays.add(i)
                ans[i] = 1
        return ans

File: problems/1488_avoid_flood_in_the_city/1488_avoid_flood_in_the_city.java
class Solution {
    public int[] avoidFlood(int[] rains) {
        int n = rains.length;
        int[] ans = new int[n];
        TreeSet<Integer> dryDays = new TreeSet<>();
        Map<Integer, Integer> lastRain = new HashMap<>(); // lake -> last rain day

        for (int i = 0; i < n; i++) {
            int lake = rains[i];
            if (lake > 0) {
                ans[i] = -1;
                if (lastRain.containsKey(lake)) {
                    Integer d = dryDays.higher(lastRain.get(lake));
                    if (d == null) {
                        return new int[0];
                    }
                    ans[d] = lake;
                    dryDays.remove(d);
                }
                lastRain.put(lake, i);
            } else {
                dryDays.add(i);
                ans[i] = 1;
            }
        }
        return ans;
    }
}